Document/0x07b-Testing-Network-Communication.md

```diff
@@ -227,3 +227,87 @@ Ensure that critical operations require at least one additional channel to confi
+
+##### Server certificate validation
+
+Start the analysis by testing the application's behavior while establishing secure connection. The test approach is to gradually relax security of SSL/TLS handshake negotiation and check which security mechanisms are enabled.
+
+1. Using Burp set up a proxy in WiFi settings. Make sure that no certificates are added to trust store (iOS: Settings -> General -> Profiles, Android: depends on the vendor and version of the OS) and that tools like SSL Kill Switch are deactivated. Launch your application and check if you can see the traffic in Burp. Any failures will be reported under the 'Alerts' tab. If you can see the traffic, it means that certificate validation is not performed at all! This effectively means that an active attacker can silently do MiTM against your application. If, however, you can't see any traffic and you have information about the SSL handshake failure, proceed to the next step.
+2. Install the Burp certificate, as explained in the Basic Security Testing section. If the handshake is successful and you can see the traffic in Burp, it means that certificate is validated against device's trust store, but the pinning is not performed. The risk is less significant than in the previous scenario, as two main attack scenarios at this point are misbehaving CAs and phishing attacks, as discussed in Basic Security Testing section.
+3. If executing instructions from previous step doesn't lead to traffic being proxied through Burp, it means that certificate is actually pinned and all security measures are in place. However, you still need to bypass the pinning in order to test the application. Please refer to Basic Security Testing section for more information on this.
+
+##### Client certificate validation
+
+Some applications use a two-way SSL handshake, meaning that the application verifies server's certificate and the server verifies the client's certificate. You can notice this if there is an error in Burp 'Alerts' tab indicating that client failed to negotiate connection.
+
+There are a couple of things worth noting:
+
+1. Client certificate contains private key that will be used in key exchange.
+2. Usually certificate would also need a password to use (decrypt) it.
+3. Certificate itself can be stored in the binary itself, data directory or the keychain.
+
+Most common but improper and insecure way of doing two-way handshake is to store client certificate within the application bundle and hardcode the password. This obviously does not bring much security, because all clients will share the same certificate.
+
+Second way of storing the certificate (and possibly password) is to use the keychain. Upon first login, the application should download personal certificate and store it securely in the keychain.
```

Comment:
bear in mind: a keychain in android does something else than the keychain in ios :D.